src/lib/sql/sql_translator.cpp

```diff
@@ -403,15 +410,103 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_join(const hsql::Join
   join_node->set_left_input(left_node);
   join_node->set_right_input(right_node);
-
+  _apply_predicate_expressions(join_node, LQPInputSide::Left, left);
+  _apply_predicate_expressions(join_node, LQPInputSide::Right, right);
   return join_node;
 }
 
+LQPInputSide SQLTranslator::_get_side(const std::shared_ptr<AbstractLQPNode>& left_node,
+                                      const std::shared_ptr<AbstractLQPNode>& right_node, const hsql::Expr* expr) {
+  DebugAssert(expr && expr->type == hsql::kExprColumnRef, "only applicable to columnar expressions");
+  const auto qualified = HSQLExprTranslator::to_qualified_column_name(*expr);
+  if (left_node->find_column(qualified)) {
+    return LQPInputSide::Left;
+  }
+
+  if (right_node->find_column(qualified)) {
+    return LQPInputSide::Right;
+  }
+
+  Fail("Column not found: " + qualified.as_string());
+}
+
+void SQLTranslator::_get_sides_from_operator_expressions(std::vector<const hsql::Expr*> operators,
+                                                         std::function<LQPInputSide(const hsql::Expr*)> get_side,
+                                                         std::vector<const hsql::Expr*>& both,
+                                                         std::vector<const hsql::Expr*>& left,
+                                                         std::vector<const hsql::Expr*>& right) {
+  for (const auto* op : operators) {
+    const bool left_is_column = op->expr && op->expr->type == hsql::kExprColumnRef;
+    const bool right_is_column = op->expr2 && op->expr2->type == hsql::kExprColumnRef;
+
+    if (left_is_column && right_is_column) {
+      auto left_column_side = get_side(op->expr);
+      auto right_column_side = get_side(op->expr2);
+      if (left_column_side == right_column_side) {
+        if (left_column_side == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else {
+        both.push_back(op);
+      }
+    } else {
+      if (left_is_column) {
+        if (get_side(op->expr) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else if (right_is_column) {
+        if (get_side(op->expr2) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      }
+    }
+  }
+}
+
+void SQLTranslator::_apply_predicate_expressions(const std::shared_ptr<AbstractLQPNode>& node, const LQPInputSide side,
+                                                 const std::vector<const hsql::Expr*>& conditions) {
+  if (conditions.empty()) {
+    return;
+  }
+
+  std::vector<std::shared_ptr<AbstractLQPNode>> predicates;
+  for (const auto* condition : conditions) {
+    predicates.push_back(_translate_predicate(
+        *condition, false, [&node](const auto& expr) { return HSQLExprTranslator::to_column_reference(expr, node); },
```

Comment:
Re consider the `for (const auto* condition : conditions) {` and the `for (size_t index = 0; index < predicates.size() - 1; ++index) {` loop. I think the second one break ORs, as you assume that a predicate consists of a single node, which it doesn't have to. 
Instead passing the result from _translate_predicate as the last param into the next invocation (and starting with node->input(side)) will build the correct LQP.

Please add a Test that verifies that OR predicates in outer Joins will be handled correctly. `SELECT * FROM A LEFT JOIN B WHERE A.j = B.j AND (B.a ="x" or B.b = "y");`, e.g.

Reply:
Yes, thank you. The proposed change remedies @mrks concerns on the method's impl, cuts one loop, functions correctly in case of predicates resulting in more than node, and is believed to be easier understandable.